parquet-column/src/test/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,413 @@
+
+    return true;
+  }
+
+  /**
+   * Calculate optimal size according to the number of distinct values and false positive probability.
+   * @param n: The number of distinct values.
+   * @param p: The false positive probability.
+   * @return optimal number of bits of given n and p.
+   */
+  public static int optimalNumOfBits(long n, double p) {
+    Preconditions.checkArgument((p > 0.0 && p < 1.0),
+      "FPP should be less than 1.0 and great than 0.0");
+
+    final double M = -8 * n / Math.log(1 - Math.pow(p, 1.0 / 8));
+    final double MAX = ParquetProperties.DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES << 3;
+    int numBits = (int)M;
+
+    // Handle overflow.
+    if (M > MAX || M < 0) {
+      numBits = (int)MAX;
+    }
+
+    // Get next power of 2 if bits is not power of 2.
+    if ((numBits & (numBits - 1)) != 0) {
+      numBits = Integer.highestOneBit(numBits) << 1;
+    }
+
+    // Minimum
+    if (numBits < (BYTES_PER_FILTER_BLOCK << 3)) {
+      numBits = BYTES_PER_FILTER_BLOCK << 3;
+    }
+
+    return numBits;
+  }
+
+  /**
+   * used to decide if we want to work to the next page
+   * @return Bytes buffered of bloom filter.
+   */
+  public long getBufferedSize() {
+    return bitset.length;
+  }
+
+  /**
+   * Compute hash for int value by using its plain encoding result.
+   * @param value the value to hash
+   * @return hash result
+   */
+  public long hash(int value) {
+    ByteBuffer plain = ByteBuffer.allocate(Integer.SIZE/Byte.SIZE);
+    plain.order(ByteOrder.LITTLE_ENDIAN).putInt(value);
+    return hashFunction.hashBytes(plain.array()).asLong();
+  }
+
+  /**
+   * Compute hash for long value by using its plain encoding result.
+   * @param value the value to hash
+   * @return hash result
+   */
+  public long hash(long value) {
+    ByteBuffer plain = ByteBuffer.allocate(Long.SIZE/Byte.SIZE);
+    plain.order(ByteOrder.LITTLE_ENDIAN).putLong(value);
+    return hashFunction.hashBytes(plain.array()).asLong();
+  }
+
+  /**
+   * Compute hash for double value by using its plain encoding result.
+   * @param value the value to hash
+   * @return hash result
+   */
+  public long hash(double value) {
+    ByteBuffer plain = ByteBuffer.allocate(Double.SIZE/Byte.SIZE);
+    plain.order(ByteOrder.LITTLE_ENDIAN).putDouble(value);
+    return hashFunction.hashBytes(plain.array()).asLong();
+  }
+
+  /**
+   * Compute hash for float value by using its plain encoding result.
+   * @param value the value to hash
+   * @return hash result
+   */
+  public long hash(float value) {
+    ByteBuffer plain = ByteBuffer.allocate(Float.SIZE/Byte.SIZE);
+    plain.order(ByteOrder.LITTLE_ENDIAN).putFloat(value);
+    return hashFunction.hashBytes(plain.array()).asLong();
+  }
+
+  /**
+   * Compute hash for Binary value by using its plain encoding result.
+   * @param value the value to hash
+   * @return hash result
+   */
+  public long hash(Binary value) {
+      return hashFunction.hashBytes(value.toByteBuffer().array()).asLong();
```

Comment:
I think this is still a open since we agreed to use plain encoding input. In parquet thrift definition,  BYTE_ARRAY is encoded with length, and FIXED_LEN_BYTE_ARRAY is encoded without length.  In PlainValuesWriter.java, it stored length for Binary.

Do we need to still use plain encoding input for Binary?  @rdblue @jbapple-cloudera

Reply:
We don't need to include the length.

Also, this shouldn't use `.array()` without checking whether the buffer is direct or on heap, and without supplying an offset because the ByteBuffer may point to the middle of its backing array. Here's what I'm using in another project:

```java
    // MURMUR3 is the configured hash function from guava; this uses the 32-bit variant
    public int hash(ByteBuffer value) {
      if (value.hasArray()) {
        return MURMUR3.hashBytes(value.array(),
            value.arrayOffset() + value.position(),
            value.arrayOffset() + value.remaining()).asInt();
      } else {
        byte[] copy = new byte[value.remaining()];
        value.get(copy);
        return MURMUR3.hashBytes(copy).asInt();
      }
    }
```

There should also be test cases for the following:

* Direct byte buffer
* Direct byte buffer with non-zero `position`
* Heap byte buffer
* Heap byte buffer with non-zero `arrayOffset`
* Heap byte buffer with non-zero `position`

Reply:
The ByteBuffer is from Binary API which already consider things you mentioned.

The same as (getBytes/getUnsafeBytes[no copy]).  Looks like I can use getUnsafeBytes to avoid copy, since it doesn't change any value.

Reply:
This should use the ByteBuffer and add tests for the cases that I listed. ByteBuffers will require fewer copies. Otherwise, this will need to be rewritten before actually using it to produce bloom filters.

Reply:
I understand ByteBuffer should avoid copies, but Murmur3 from guava 20.0 only support byte[] parameter.  We can change to use ByteBuffer when update to guava23.0.

The difference from your reference code is here we use Binary as input parameter, and we should trust the APIs from Binary class, shouldn't we?  If it accept a ByteBuffer parameter, we should add tests you list accordingly.

In addition, in you reference code, why you add arrayOffset to remaining? remaining should be (limit -position) and irrelevant with arraryOffset, otherwise the length changes, am I wrong?

Reply:
Trusting an API for correctness and using the most efficient method from that API are two different things. If you use `toByteArray`, your code will often copy data it doesn't need to, which is why it will need to be replaced before actually using this. If Guava doesn't support ByteBuffer, is there another implementation that does? Can we update Guava?

Reply:
Yes, we can update Guava to 23.0, it needs some minor changes to fix build issues.  Please have a look the latest commit.  

I also update maven-shade-plugin version since it show ArrayIndexOutOfBoundsException when creating shaded jar for "com/google/common/cache/LocalCache$EntrySet.class".